File: .claude/skills/kitten/scripts/kitten-fetch.js
#!/usr/bin/env node
/**
 * kitten-fetch.js
 *
 * Fetches any file from the kitten GitHub repo using the GitHub Contents API.
 * Reads repo and branch from config.json.
 * GITHUB_TOKEN must be set in .env at the project root. No fallbacks.
 *
 * Usage:
 *   node kitten-fetch.js <file-path> [branch]
 *   node kitten-fetch.js references/kitten/stack.md
 *   node kitten-fetch.js workflows/identity.md dev
 *
 * Output:
 *   Decoded file content printed to stdout.
 */

const fs = require('fs');
const path = require('path');
const https = require('https');

// --- Config ---

const configPath = path.join(__dirname, '..', 'config.json');
const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));

const repo = config.repo;
const defaultBranch = config.branch || 'main';

// --- Token — strictly from .env at project root. No fallbacks. ---

const envPath = path.join(__dirname, '..', '..', '..', '..', '.env');

if (!fs.existsSync(envPath)) {
  console.error('Error: .env file not found at project root.');
  process.exit(1);
}

const envVars = {};
fs.readFileSync(envPath, 'utf8')
  .split('\n')
  .forEach((line) => {
    const [key, ...rest] = line.split('=');
    if (key && rest.length) envVars[key.trim()] = rest.join('=').trim();
  });

const token = envVars['GITHUB_TOKEN'];

if (!token) {
  console.error('Error: GITHUB_TOKEN is not defined in .env.');
  process.exit(1);
}

// --- Args ---

const filePath = process.argv[2];
const branch = process.argv[3] || defaultBranch;

if (!filePath) {
  console.error('Usage: node kitten-fetch.js <file-path> [branch]');
  console.error('Example: node kitten-fetch.js references/kitten/stack.md');
  process.exit(1);
}

// --- Derive API URL ---

const repoPath = repo.replace('https://github.com/', '');
const apiUrl = `https://api.github.com/repos/${repoPath}/contents/${filePath}?ref=${branch}`;

// --- Fetch ---

function get(url) {
  return new Promise((resolve, reject) => {
    const options = {
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: 'application/vnd.github.v3+json',
        'User-Agent': 'kitten-fetch',
      },
    };

    https.get(url, options, (res) => {
      let data = '';
      res.on('data', (chunk) => (data += chunk));
      res.on('end', () => {
        if (res.statusCode === 200) {
          resolve(data);
        } else {
          reject(new Error(`HTTP ${res.statusCode}: ${data}`));
        }
      });
    }).on('error', reject);
  });
}

(async () => {
  try {
    const raw = await get(apiUrl);
    const json = JSON.parse(raw);

    if (json.encoding !== 'base64' || !json.content) {
      throw new Error('Unexpected response format from GitHub API.');
    }

    const content = Buffer.from(json.content, 'base64').toString('utf8');
    process.stdout.write(content);
  } catch (err) {
    console.error(`kitten-fetch error: ${err.message}`);
    process.exit(1);
  }
})();

File: .claude/skills/kitten-bot/scripts/kitten_fetch.py
#!/usr/bin/env python3
"""
kitten_fetch.py

Fetches any file from the kitten-bot GitHub repo using the GitHub Contents API.
Reads repo and branch from config.json (in the skill root).
GITHUB_TOKEN must be set in .env at the project root. No fallbacks.

Usage (run from skill root):
  python -m scripts.kitten_fetch <file-path> [branch]
  python -m scripts.kitten_fetch references/kitten/stack.md
  python -m scripts.kitten_fetch agents/identity.md dev

Output:
  Decoded file content printed to stdout.
"""

import sys
import json
import base64
import ssl
import urllib.request
from pathlib import Path


def ssl_context() -> ssl.SSLContext:
    # Try platform cert bundles in order — Python.org builds don't bundle certs
    candidates = [
        "/etc/ssl/cert.pem",                          # macOS + some Linux
        "/etc/ssl/certs/ca-certificates.crt",         # Debian/Ubuntu
        "/etc/pki/tls/certs/ca-bundle.crt",           # RHEL/Fedora
        "/etc/ssl/ca-bundle.pem",                     # OpenSUSE
    ]
    for path in candidates:
        if Path(path).exists():
            return ssl.create_default_context(cafile=path)
    # Fall back to default (works when Python was installed with certs)
    return ssl.create_default_context()

# --- Paths (cwd = skill root because caller does: cd {skill_dir} && python -m ...) ---

skill_dir = Path.cwd()
config_path = skill_dir / "config.json"

# Walk up from skill_dir to find .env
def find_env(start: Path) -> Path | None:
    for parent in [start, *start.parents]:
        candidate = parent / ".env"
        if candidate.exists():
            return candidate
    return None

# --- Config ---

if not config_path.exists():
    print(f"Error: config.json not found at {config_path}", file=sys.stderr)
    sys.exit(1)

with open(config_path) as f:
    config = json.load(f)

repo = config["repo"]
default_branch = config.get("branch", "main")

# --- Token — strictly from .env. No fallbacks. ---

env_path = find_env(skill_dir)

if not env_path:
    print("Error: .env file not found in skill directory or any parent.", file=sys.stderr)
    sys.exit(1)

env_vars = {}
for line in env_path.read_text().splitlines():
    if "=" in line and not line.startswith("#"):
        key, _, value = line.partition("=")
        env_vars[key.strip()] = value.strip()

token = env_vars.get("GITHUB_TOKEN")

if not token:
    print("Error: GITHUB_TOKEN is not defined in .env.", file=sys.stderr)
    sys.exit(1)

# --- Args ---

if len(sys.argv) < 2:
    print("Usage: python -m scripts.kitten_fetch <file-path> [branch]", file=sys.stderr)
    print("Example: python -m scripts.kitten_fetch references/kitten/stack.md", file=sys.stderr)
    sys.exit(1)

file_path = sys.argv[1]
branch = sys.argv[2] if len(sys.argv) > 2 else default_branch

# --- Source repo detection: read local files when running inside kitten source ---

def find_repo_root(start: Path) -> "Path | None":
    for parent in [start, *start.parents]:
        if (parent / ".git").is_dir():
            return parent
    return None

def is_source_repo(root: Path) -> bool:
    git_config = root / ".git" / "config"
    if not git_config.exists():
        return False
    return "abappi19/kitten" in git_config.read_text()

repo_root = find_repo_root(skill_dir)
if repo_root and is_source_repo(repo_root):
    local_file = repo_root / file_path
    if local_file.exists():
        sys.stdout.write(local_file.read_text())
        sys.exit(0)
    else:
        print(f"kitten-fetch: local file not found at {local_file}", file=sys.stderr)
        sys.exit(1)

# --- Fetch ---

repo_path = repo.replace("https://github.com/", "")
api_url = f"https://api.github.com/repos/{repo_path}/contents/{file_path}?ref={branch}"

req = urllib.request.Request(
    api_url,
    headers={
        "Authorization": f"Bearer {token}",
        "Accept": "application/vnd.github.v3+json",
        "User-Agent": "kitten-fetch",
    },
)

try:
    with urllib.request.urlopen(req, context=ssl_context()) as res:
        data = json.loads(res.read())
except ssl.SSLCertVerificationError:
    print(
        "kitten-fetch error: SSL certificate verification failed.\n"
        "Fix: run the certificate installer for your Python version.\n"
        "  macOS:  open '/Applications/Python 3.x/Install Certificates.command'\n"
        "  Linux:  sudo apt install ca-certificates  (or equivalent for your distro)\n"
        "  Any:    pip install certifi",
        file=sys.stderr,
    )
    sys.exit(1)
except urllib.error.HTTPError as e:
    print(f"kitten-fetch error: HTTP {e.code}: {e.read().decode()}", file=sys.stderr)
    sys.exit(1)
except urllib.error.URLError as e:
    print(f"kitten-fetch error: {e.reason}", file=sys.stderr)
    sys.exit(1)

if data.get("encoding") != "base64" or not data.get("content"):
    print("kitten-fetch error: Unexpected response format from GitHub API.", file=sys.stderr)
    sys.exit(1)

content = base64.b64decode(data["content"]).decode("utf-8")
sys.stdout.write(content)
